File: Utils/src/main/java/org/tobi29/scapes/engine/utils/io/tag/TagStructureJSON.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.utils.io.tag;

import org.tobi29.scapes.engine.utils.io.ReadableByteStream;
import org.tobi29.scapes.engine.utils.io.WritableByteStream;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class TagStructureJSON {
    public static void write(TagStructure tagStructure, OutputStream streamOut)
            throws IOException {
        tagStructure.write(new TagStructureWriterJSON(streamOut));
    }

    public static TagStructure read(InputStream streamIn) throws IOException {
        return read(new TagStructure(), streamIn);
    }

    public static TagStructure read(TagStructure tagStructure,
            InputStream streamIn) throws IOException {
        tagStructure.read(new TagStructureReaderJSON(streamIn));
        return tagStructure;
    }

    public static void write(TagStructure tagStructure,
            WritableByteStream stream) throws IOException {
        tagStructure.write(new TagStructureWriterJSON(stream));
    }

    public static TagStructure read(ReadableByteStream stream)
            throws IOException {
        return read(new TagStructure(), stream);
    }

    public static TagStructure read(TagStructure tagStructure,
            ReadableByteStream stream) throws IOException {
        tagStructure.read(new TagStructureReaderJSON(stream));
        return tagStructure;
    }
}

File: tag-json/src/commonMain/kotlin/TagStructureJSON.kt
/*
 * Copyright 2012-2018 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.io.tag.json

import org.tobi29.io.EndOfStreamException
import org.tobi29.io.tag.TagMap
import org.tobi29.io.tag.writeContents
import org.tobi29.stdex.Readable

fun readJSON(stream: Readable): TagMap = readJSON { stream.read() }

fun readJSON(
    str: String,
    offset: Int = 0,
    size: Int = str.length - offset
): TagMap {
    val end = offset + size
    var i = offset
    return readJSON {
        if (i >= end) throw EndOfStreamException()
        str[i++]
    }
}

fun TagMap.writeJSON(stream: Appendable, pretty: Boolean = true) {
    if (pretty) {
        TagStructureWriterPrettyJSON(stream).let { writer ->
            writer.begin(this)
            writeContents(writer)
            writer.end()
        }
    } else {
        TagStructureWriterMiniJSON(stream).let { writer ->
            writer.begin(this)
            writeContents(writer)
            writer.end()
        }
    }
}
